Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+# Standalone GC Eventing Design
+
+Author: Sean Gillespie (@swgillespie) - 2017
+
+This document aims to provide a specification for how a standalone GC fires
+events that can be collected by trace collectors. Such a feature is highly desirable
+for a standalone GC since it is the primary way that is used to reason about
+GC performance. Since a standalone GC is not permitted to link against the rest of the runtime, all
+communication betwene the runtime and the GC most pass through dynamically-dispatched interfaces.
+
+## Definitions
+
+* An **event** is some unit of information that the runtime can issue if requested. In general,
+this is used for lightweight tracing. Managed code can issue events using
+`System.Diagnostics.Tracing.EventSource`. Native code (i.e. the runtime) issues events by calling macros
+that delegate the issuing of events to autogenerated code that is generated to interface with the
+underlying event implementation. Events are only issued if they are turned on; the mechanism by which
+events are turned on is not in the scope of this document
+* The **payload** of an event is some amount of data that is delivered with the event itself. Its size
+may be variable. Most events that are fired by the runtime have a schema (predefined layout), but it
+is not a requirement.
+
+## Goals
+
+The goal of this document is to describe a system that allows for the efficient firing of
+performance events by a standalone GC. This system must have two properties in order to be
+acceptable:
+
+1. It must be *efficient* to query whether or not a particular event is turned on. It is not acceptable
+to perform an indirection (i.e. cross the GC/EE interface boundary) in order to get this information.
+2. The cost of firing an event by a standalone GC should be comparable to the cost of firing an event
+without using a standalone GC.
+
```

Comment:
being able to add new events should also be a goal.

Reply:
sure! I'll call that out explicitly here.